src/base/utils/foreignapps.cpp

```diff
@@ -0,0 +1,126 @@
+ * and distribute the linked executables. You must obey the GNU General Public
+ * License in all respects for all of the code used other than "OpenSSL".  If you
+ * modify file(s), you may extend this exception to your version of the file(s),
+ * but you are not obligated to do so. If you do not wish to do so, delete this
+ * exception statement from your version.
+ */
+
+#include "foreignapps.h"
+
+#include <QCoreApplication>
+#include <QProcess>
+#include <QStringList>
+
+#include "base/logger.h"
+#include "base/utils/version.h"
+
+struct PythonInfo
+{
+    using Version = Utils::Version<qint8, 3, 1>;
+
+    bool isEmpty() const
+    {
+        return (executableName.isEmpty() || (version == Version {}));
+    }
+
+    QString executableName;
+    Version version {};
+};
+
+namespace
+{
+    PythonInfo pythonInfo()
+    {
+        static PythonInfo pyInfo;
+        if (pyInfo.isEmpty()) {
+            const auto testPythonInstallation = [](const QString &exeName) -> bool
+            {
+                QProcess proc;
+                proc.start(exeName, {"--version"}, QIODevice::ReadOnly);
+                if (proc.waitForFinished() && (proc.exitCode() == QProcess::NormalExit)) {
+                    QByteArray procOutput = proc.readAllStandardOutput();
+                    if (procOutput.isEmpty())
+                        procOutput = proc.readAllStandardError();
+                    procOutput = procOutput.simplified();
+
+                    // Software 'Anaconda' installs its own python interpreter
+                    // and `python --version` returns a string like this:
+                    // "Python 3.4.3 :: Anaconda 2.3.0 (64-bit)"
+                    const QList<QByteArray> outputSplit = procOutput.split(' ');
+                    if (outputSplit.size() <= 1)
+                        return false;
+
+                    pyInfo.executableName = exeName;
+                    pyInfo.version = outputSplit.at(1);
+                    LogMsg(QCoreApplication::translate("python", "Python detected, version: %1").arg(pyInfo.version), Log::INFO);
+                    return true;
+                }
+
+                return false;
+            };
+
+#if defined(Q_OS_UNIX)
+            // On Unix-Like Systems python2 and python3 should always exist
+            // https://legacy.python.org/dev/peps/pep-0394/
+            if (testPythonInstallation("python3"))
+                return pyInfo;
+            if (testPythonInstallation("python2"))
+                return pyInfo;
+#endif
+            // Look for "python" in Windows and in UNIX if "python2" and "python3" are
+            // not detected.
+            if (testPythonInstallation("python"))
+                return pyInfo;
+
+            LogMsg(QCoreApplication::translate("python", "Python not detected"), Log::INFO);
+        }
+
+        return pyInfo;
+    }
+}
+
+/**
+ * Returns the major version number from `python --version`.
+ */
+int Utils::ForeignApps::Python::majorVersion()
```

Comment:
Why not just write one function that returns the PythonInfo, instead of these three?

Reply:
https://github.com/qbittorrent/qBittorrent/pull/9113#discussion_r196764590
https://github.com/qbittorrent/qBittorrent/pull/9113#discussion_r196765375

Reply:
I don't see how they contradict...

Reply:
But I don't want to rewrite all usages of the existing functions to the new one....

Reply:
>But I don't want to rewrite all usages of the existing functions to the new one....

But you're changing them anyway!
What do you care if you change it this way: `Utils::Misc::pythonExecutable()` => `Utils::ForeignApps::Python::executableName()` or the other way: `Utils::Misc::pythonExecutable()` => `Utils::ForeignApps::Python::info().executableName`?
In addition, we can save one level of namespaces:
```c++
namespace Utils
{
    namespace ForeignApps
    {
        PythonInfo pythonInfo();
        App1Info app1Info();
        App2Info app2Info();
        // and so on...
    }
}
```

Reply:
and what do you suggest about `fullVersion()`?

Reply:
I don't really see why you need it. Why don't you take @zeule suggestion (maybe not exactly as he suggested) and just use Version elements?
```c++
        Utils::ForeignApps::PythonInfo pythonInfo = Utils::ForeignApps::pythonInfo();

        if ((pythonInfo.version.majorNumber() == 2) || (pythonInfo.version.majorNumber() == 3)) {
            if (((pythonInfo.version.majorNumber() == 2) && (pythonInfo.version.minorNumber() < 7))
                || ((pythonInfo.version.majorNumber() == 2) && (pythonInfo.version.minorNumber() == 7) && (pythonInfo.version.revisionNumber() < 9))
                || ((pythonInfo.version.majorNumber() == 3) && (pythonInfo.version.minorNumber() < 3))) {
```
Or even you can use `operator[]` instead of `xxxNumber()`.